File: 1.semester/algorithmAndDesign/assigments/kattis/dynamicProgramming/canonicalCoinSystem/Program.cpp
#include <fstream>
#include <iostream>
#include <istream>
#include <string>
#include <vector>

std::vector<int> readData(std::istream &inputStream) {
    std::string line;
    std::getline(inputStream, line);

    int n = std::stoi(line);
    std::vector<int> valueVector{};

    for (int i = 0; i < n; i++) {
        int a = 0;
        inputStream >> a;
        valueVector.push_back(a);
    }

    return valueVector;
}

std::vector<int> getInputData(int argc, char **argv) {
    if (argc > 1) {
        std::ifstream ifile(argv[1]);
        return readData(ifile);
    } else {
        return readData(std::cin);
    }
}

bool greedyApproach(std::vector<int> &coins, int amount, int stopAt) {
    int counter = 0;
    
    while (amount > 0) {
        for (int i = coins.size() - 1; i >= 0; i--) {
            int coin = coins[i];
            if (coin <= amount) {
                int divided = amount / coin;

                amount -= coin * divided;
                counter += divided;

                if (counter > stopAt) {
                    return false;
                }

                break;
            }
        }
    }
    return true;
}

std::vector<int> dynamicApproach(std::vector<int> &coins, int amount) {
    std::vector<int> M(amount + 1, 0);

    for (int coin : coins) {
        for (int currentAmount = coin; currentAmount < amount + 1; currentAmount++) {
            if (coin <= currentAmount) {
                int leftover = currentAmount - coin;
                int coinsUsedForLeftover = M[leftover];

                if (M[currentAmount] == 0) {
                    M[currentAmount] = coinsUsedForLeftover + 1;
                } else {
                    M[currentAmount] = std::min(coinsUsedForLeftover + 1, M[currentAmount]);
                }
            }
        }
    }
    return M;
}

void solve(std::vector<int> &coins) {
    int upperBound = coins[coins.size() - 1] + coins[coins.size() - 2] + 1;
    int lowerBound = coins[coins.size() - 3];

    auto dynamicResult = dynamicApproach(coins, upperBound);

    for (int i = upperBound; i >= lowerBound; i--) {
        if (!greedyApproach(coins, i, dynamicResult[i])) {
            std::cout << "non-canonical" << std::endl;
            return;
        }
    }

    std::cout << "canonical" << std::endl;
}

int main(int argc, char **argv) {
    auto coins = getInputData(argc, argv);
    solve(coins);
}

File: 1.semester/algorithmAndDesign/assigments/kattis/dynamicProgramming/canonicalCoinSystem/program.py
import fileinput
import sys


def main():
    d, coins = get_formatted_data()

    upper_bound = coins[-1] + coins[-2]
    lower_bound = coins[-3]

    dynamic_res = dynamic_approach(coins, upper_bound)

    for i in reversed(range(lower_bound, upper_bound)):
        if(greedy_approach(coins, i) != dynamic_res[i]):
            print("non-canonical") 
            exit()
    
    print("canonical")
  


def get_formatted_data():
    data_input = create_data_input()

    d = int(data_input.readline())
    coins = [int(coin) for coin in data_input.readline().strip().split(" ")]

    return d, coins


def greedy_approach(coins: list, amount: int) -> int:
    counter = 0
    while(amount > 0):
        for coin in reversed(coins):
            if coin <= amount:

                divided = amount / coin
                
                amount = amount -  (coin * divided);
                counter = counter + divided;

                break
    return counter


def dynamic_approach(coins: list,  amount: int) -> list:
    M = [0] * (amount + 1)

    for coin in coins:
        for current_amount in range(amount + 1):
            if coin <= current_amount:
                leftover = current_amount - coin
                coins_used_for_leftover = M[leftover]
                if M[current_amount] == 0:
                    M[current_amount] = coins_used_for_leftover + 1
                else:
                    M[current_amount] = min(coins_used_for_leftover + 1, M[current_amount])
                

    return M


def create_data_input() -> fileinput.FileInput:
    data = None
    if len(sys.argv) > 1:
        data = fileinput.input(sys.argv[1])
    else:
        data = fileinput.input()

    return data


if __name__ == "__main__":
    main()
